sim/midas/src/main/cc/emul/vcs-harness.cc

```diff
@@ -218,40 +218,39 @@ void tick(vc_handle reset,
           vc_handle mem_3_b_bits_resp,
           vc_handle mem_3_b_bits_id) {
   try {
-    mmio_t *m, *d;
-    assert(m = dynamic_cast<mmio_t *>(master.get()));
-    assert(d = dynamic_cast<mmio_t *>(dma.get()));
```

Comment:
It helps if I read the whole description of the PR carefully before going to do the review...   I see now where you say:
> The linkage of mmio_t mm_t instances between the driver context and the vcs/verilator context is a recurring problem. These are both global and unique_ptrs -- which seems like a gross misuse of unique_ptr. In SiFive's fork, i've switched these to be raw pointers because i was seeing instances of the runtime double-freeing the underlying data structure.
> 
> I don't really see an easy way, short of re-engineeering how metasimulation is implemented (which is on the long list of things to do), to avoid using globals to share references to mmio_t and mm_ts. So it seems to me that either switching them to being shared_ptrs or raw pointers freed when the driver is torn down are the two best options. I'm open to alternatives.

Reply:
Ok, after looking at how `master` and `dma` are [declared](https://github.com/firesim/firesim/blob/f9b0f1cf40e56942c6e5831eaacd15d7d94c8bea/sim/midas/src/main/cc/simif_emul.cc#L18-L19) and [initialized](https://github.com/firesim/firesim/blob/f9b0f1cf40e56942c6e5831eaacd15d7d94c8bea/sim/midas/src/main/cc/emul/mmio.cc#L96-L97). If I ignore whether it is good design to use globals for those pointers, I see that your change is effectively dropping the assertion that those pointers have been initialized  before this `tick()` is called.

At a minimum we should put the assertions back in:
```
assert(master.get() != nullptr)
assert(dma.get() != nullptr)
```
(if we change them to raw pointers, we should not call `.get()` on them but we still should have the assertions.

Reply:
From our offline discussion, I agree that we should just use raw pointers here until we rearchitect the code.